### PR TITLE
feature: add input moderation on all chains and all agents

### PR DESCRIPTION
**Pull request by ocflowiseai**
Draft PR with work in progress for input moderation on Chains and Agents.

### Chains
- [x] OpenAPI Chain
- [x] Conversational Retrieval QA Chain
- [x] Multi Prompt Chain
- [x] Multi Retrieval QA Chain
- [x] Retrieval QA Chain
- [x] Sql Database Chain
- [x] Vectara QA Chain
- [x] VectorDB QA Chain

### Agents
- [x] Airtable Agent
- [x] AutoGPT
- [x] BabyAGI
- [x] Conversational Agent
- [x] Conversational Retrieval Agent
- [x] CSV Agent
- [x] ReAct Agent for Chat Models
- [x] ReAct Agent for LLMs
- [x] OpenAI Assistant
- [x] OpenAI Function Agent
- [x] XML Agent

Relates to: https://github.com/FlowiseAI/Flowise/pull/1646

### CODE REVIEW

File: packages/components/nodes/chains/VectaraChain/VectaraChain.ts [hunk 1]
```diff
@@ -252,6 +262,18 @@ class VectaraChain_Chains implements INode {
...
         const mmrEnabled = vectaraFilter?.mmrConfig?.enabled
 
+        const moderations = nodeData.inputs?.inputModeration as Moderation[]
+        if (moderations && moderations.length > 0) {
+            try {
+                // Use the output of the moderation chain as input for the Vectara chain
+                input = await checkInputs(moderations, input)
+            } catch (e) {
+                await new Promise((resolve) => setTimeout(resolve, 500))
+                //streamResponse(options.socketIO && options.socketIOClientId, e.message, options.socketIO, options.socketIOClientId)
```
Comment (ocflowiseai) [inline packages/components/nodes/chains/VectaraChain/VectaraChain.ts:272]: Should Vectara QA Chain have streaming?

Comment (HenryHengZJ) [reply packages/components/nodes/chains/VectaraChain/VectaraChain.ts:272]: not it shouldnt

**Review by HenryHengZJ — APPROVED** [review]

File: packages/components/nodes/agents/OpenAIFunctionAgent/OpenAIFunctionAgent.ts [hunk 2]
```diff
@@ -67,6 +77,19 @@ class OpenAIFunctionAgent_Agents implements INode {
...
         const memory = nodeData.inputs?.memory as FlowiseMemory
+        const moderations = nodeData.inputs?.inputModeration as Moderation[]
+
+        if (moderations && moderations.length > 0) {
+            try {
+                // Use the output of the moderation chain as input for the OpenAI Function Agent
+                input = await checkInputs(moderations, input)
+            } catch (e) {
+                await new Promise((resolve) => setTimeout(resolve, 500))
+                //streamResponse(options.socketIO && options.socketIOClientId, e.message, options.socketIO, options.socketIOClientId)
```
Comment (HenryHengZJ) [inline packages/components/nodes/agents/OpenAIFunctionAgent/OpenAIFunctionAgent.ts:88]: @ocflowiseai we should enable the `streamResponse` for some of the agents/chains.

For full list of which chains/agents are available for streaming, refer [here](https://github.com/FlowiseAI/Flowise/blob/main/packages/server/src/utils/index.ts#L915)

Comment (ocflowiseai) [reply packages/components/nodes/agents/OpenAIFunctionAgent/OpenAIFunctionAgent.ts:88]: Sure thing! I will open a separate PR for that.

**Pull request merged**